src/python/pants/source/source_root.py

```diff
@@ -361,6 +362,24 @@ def _do_add_pattern(self, pattern, langs, category):
     node.category = category
     node.is_terminal = True
 
+  def traverse(self) -> Set[Tuple[str, Tuple[str, ...]]]:
```

Comment:
It's not clear that this method needs to be on the `Trie`: it feels like it could just be on the `Config`, which can generate all of these patterns directly?

Reply:
I think it's neater to have it on the trie, since it's the "front door" that the rest of the code uses to access source roots, whereas config is just one (albeit the main) source of source roots.

We do actually add source roots directly, not via config, in some cases: 

https://github.com/pantsbuild/pants/blob/6f7a34bcd58df97f7e50c3f018305a882fee6bcc/src/python/pants/goal/context.py#L291

Reply:
> We do actually add source roots directly, not via config, in some cases:

That will only have an effect on v1... synthetic targets aren't a thing in v2.

Reply:
True. Nonetheless it seems like the Trie is the right place for this, at least while that's what code uses to query source roots.